horovod/run/run_func.py

```diff
@@ -0,0 +1,220 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import print_function
+import cloudpickle
+import tempfile
+import os
+import sys
+import six
+import subprocess
+import collections
+import textwrap
+import time
+
+from horovod.run.util import network
+from horovod.run.common.util import safe_shell_exec
+
+_TAIL_LINES_TO_KEEP = 100
+
+_LOCAL_PICKLED_RESULT_FILENAME = "local_result.pkl"
+_PICKLED_RESULT_FILENAME = "result.pkl"
+_PICKLED_PROC_FN_FILENAME = "proc_fn.pkl"
+_PROC_LAUNCHER_FILENAME = "launch.sh"
+
+
+def run_func(
+        fn,
+        args=None,
+        kwargs=None,
+        num_proc=None,
+        host=None,
+        ssh_port=None,
+        disable_cache=False,
+        start_timeout=None,
+        env=None,
+        verbose=False,
+        driver_host=None):
+    """
+    Run horovod inside python program.
+    Run `num_proc` processes executing `fn` on specified hosts.
+
+    :param fn: Function to run.
+    :param args: Arguments to pass to `fn`.
+    :param kwargs: Keyword arguments to pass to `fn`.
+    :param num_proc: Number of Horovod processes. Default to be number of all host slots.
+    :param host: To specify the list of host names as well as the
+                 number of available slots on each host for
+                 training processes using the following format:
+                 <hostname>:<number of slots>,... .
+                 E.g., host1:2,host2:4,host3:1 indicates that 2 processes can run on
+                 host1, 4 processes on host2, and 1 process on host3.
+    :param ssh_port: SSH port on all the hosts.
+    :param disable_cache: The same argument with horovodrun command.
+    :param start_timeout: Horovodrun has to perform all the checks and
+                          start the processes before the specified
+                          timeout. The default value is 30 seconds.
+                          Alternatively, The environment variable
+                          HOROVOD_START_TIMEOUT can also be used to
+                          specify the initialization timeout.
+    :param env: Environment dictionary to use in Horovod run.  Defaults to `os.environ`.
+    :param verbose: If this flag is set, extra messages will be printed.
+    :param driver_host: The host of node `run_func` running on.
+    :return: The returned value of rank 0 process.
```

Comment:
@alsrgv Do you prefer collecting all results or only from rank #0?

Reply:
Update to be collecting all processes results.